packages/flutter/lib/src/cupertino/date_picker.dart

```diff
@@ -0,0 +1,358 @@
+  /// and must not be null.
+  ///
+  /// [initialTimerDuration] defaults to 0 second and is limited from 0 second
+  /// to 23 hours 59 minutes 59 seconds. Only hour, minute, and second values
+  /// are extracted from [initialTimerDuration], so specifying other fields like
+  /// millisecond or microsecond will not affect anything.
+  ///
+  /// [minuteInterval] is the granularity of the minute spinner. Must be a factor
+  /// of 60.
+  ///
+  /// [secondInterval] is the granularity of the second spinner. Must be a factor
+  /// of 60.
+  CupertinoCountdownTimerPicker({
+    this.initialTimerDuration = const Duration(),
+    this.minuteInterval = 1,
+    this.secondInterval = 1,
+    @required this.onTimerDurationChanged,
+  }) : assert(onTimerDurationChanged != null),
+       assert(initialTimerDuration >= const Duration(seconds: 0)),
+       assert(initialTimerDuration < const Duration(days: 1)),
+       assert(minuteInterval > 0 && 60 % minuteInterval == 0),
+       assert(secondInterval > 0 && 60 % secondInterval == 0),
+       assert(initialTimerDuration.inMinutes % minuteInterval == 0),
+       assert(initialTimerDuration.inSeconds % secondInterval == 0);
+
+  /// The initial duration of the countdown timer.
+  final Duration initialTimerDuration;
+
+  /// The granularity of the minute spinner. Must be a factor of 60.
+  final int minuteInterval;
+
+  /// The granularity of the second spinner. Must be a factor of 60.
+  final int secondInterval;
+
+  /// Callback when the timer duration changes.
+  final ValueChanged<Duration> onTimerDurationChanged;
+
+  @override
+  State<StatefulWidget> createState() => _CountdownTimerState();
+}
+
+class _CountdownTimerState extends State<CupertinoCountdownTimerPicker> {
+  // The currently selected values of the picker.
+  int selectedHour;
+  int selectedMinute;
+  int selectedSecond;
+
+  // Controllers for the 3 units: hour, minute, and second.
+  FixedExtentScrollController hourController;
+  FixedExtentScrollController minuteController;
+  FixedExtentScrollController secondController;
+
+  @override
+  void initState() {
+    super.initState();
+
+    selectedHour = widget.initialTimerDuration.inHours;
+    selectedMinute = widget.initialTimerDuration.inMinutes % 60;
+    selectedSecond = widget.initialTimerDuration.inSeconds % 60;
+
+    hourController = new FixedExtentScrollController(initialItem: selectedHour);
+    minuteController = new FixedExtentScrollController(
+      initialItem: selectedMinute ~/ widget.minuteInterval);
+    secondController = new FixedExtentScrollController(
+      initialItem: selectedSecond ~/ widget.secondInterval);
+  }
+
+  // Builds a text label with customized scale factor and font weight.
+  Widget _buildLabel(String text) {
+    return new Text(
+      text,
+      textScaleFactor: 0.8,
+      style: const TextStyle(fontWeight: FontWeight.w600),
+    );
+  }
+
+  Widget _buildHourPicker(BuildContext context) {
+    final int textDirectionFactor =
```

Comment:
calling this isn't particularly expensive but you might as well do it once in didChangeDependencies() to avoid code duplication and maintenance overhead